File: servers/java/vertx/app/src/main/java/reis/polyglot/vertx/server/lib/RequestSchemaFactory.java
package reis.polyglot.vertx.server.lib;

import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import java.util.Arrays;

public class RequestSchemaFactory {
    private static final JsonObject defaultSchema = new JsonObject()
        .put("type", "object")
        .put("properties", new JsonObject())
        .put("required", new JsonArray());
    private static final JsonObject defaultQuerySchema = new JsonObject()
        .put("type", "object")
        .put("properties", new JsonObject()
            .put("fields", new JsonObject()
                .put("type", "string")
            )
            .put("page", new JsonObject()
                .put("type", "number")
                .put("minimum", 0)
                .put("default", 0)
            )
            .put("pageSize", new JsonObject()
                .put("type", "number")
                .put("minimum", 1)
                .put("default", 50)
            )
            .put("sort", new JsonObject()
                .put("type", "object")
            )
            .put("search", new JsonObject()
                .put("type", "string")
            )
        )
        .put("required", new JsonArray());

    public static JsonObject createBodySchema(JsonObject bodySchema) {
        return createRequestSchema(defaultSchema, bodySchema, defaultSchema);
    }

    public static JsonObject createBodySchema(JsonObject bodySchema, JsonObject paramsSchema) {
        return createRequestSchema(paramsSchema, bodySchema, defaultSchema);
    }

    public static JsonObject createQuerySchema(JsonObject querySchema) {
        return createRequestSchema(defaultSchema, defaultSchema, querySchema);
    }

    public static JsonObject createQuerySchema(JsonObject querySchema, JsonObject paramsSchema) {
        return createRequestSchema(paramsSchema, defaultSchema, querySchema);
    }

    public static JsonObject createParamsSchema(JsonObject paramsSchema) {
        return createRequestSchema(paramsSchema, defaultSchema, defaultSchema);
    }

    public static JsonObject createRequestSchema(
        JsonObject paramsSchema,
        JsonObject bodySchema,
        JsonObject querySchema
    ) {
        JsonObject mergedQuerySchema = new JsonObject().mergeIn(defaultQuerySchema).mergeIn(querySchema, true);
        return new JsonObject()
            .put("type", "object")
            .put("required", new JsonArray(Arrays.asList("body", "params", "query")))
            .put("properties", new JsonObject()
                .put("body", bodySchema)
                .put("params", paramsSchema)
                .put("query", mergedQuerySchema)
            );
    }
}

File: servers/go/gin/src/api/v1/lib/validation/request_schema_factory.go
package validation

import (
	"github.com/gin-gonic/gin"
	"github.com/xeipuuv/gojsonschema"
)

func CreateBodySchema(bodySchema gin.H) gojsonschema.Schema {
	return createBaseSchema(bodySchema, gin.H{}, gin.H{})
}

func CreateBodySchemaWithParams(bodySchema gin.H, paramsSchema gin.H) gojsonschema.Schema {
	return createBaseSchema(bodySchema, gin.H{}, paramsSchema)
}

func CreateQuerySchema(querySchema gin.H) gojsonschema.Schema {
	return createBaseSchema(gin.H{}, querySchema, gin.H{})
}

func CreateQuerySchemaWithParams(querySchema gin.H, paramsSchema gin.H) gojsonschema.Schema {
	return createBaseSchema(gin.H{}, querySchema, paramsSchema)
}

func createBaseSchema(bodySchema gin.H, querySchema gin.H, paramsSchema gin.H) gojsonschema.Schema {
	baseSchema := gin.H{
		"type": "object",
		"properties": gin.H{
			"body":         bodySchema,
			"querySchema":  querySchema,
			"paramsSchema": paramsSchema,
		},
		"required": [3]string{"body", "query", "params"},
	}
	loader := gojsonschema.NewGoLoader(baseSchema)
	schema, _ := gojsonschema.NewSchema(loader)
	return *schema
}
